services/repository/src/main/java/org/sagebionetworks/repo/web/OAuthScopeInterceptor.java

```diff
@@ -0,0 +1,137 @@
+	
+	@Autowired
+	private OIDCTokenHelper oidcTokenHelper;
+	
+	public static boolean hasUserIdParameterOrAccessTokenHeader(HandlerMethod handlerMethod) {
+		for (MethodParameter methodParameter : handlerMethod.getMethodParameters()) {
+			RequestParam requestParam = methodParameter.getParameterAnnotation(RequestParam.class);
+			if (requestParam!=null && requestParam.value().equals(AuthorizationConstants.USER_ID_PARAM)) {
+				return true;
+			}
+			RequestHeader requestHeader = methodParameter.getParameterAnnotation(RequestHeader.class);
+			if (requestHeader!=null && requestHeader.value().equals(SYNAPSE_AUTHORIZATION_HEADER_NAME)) {
+				return true;
+			}
+		}
+		return false;
+	}
+	
+	public static boolean isAnonymous(HttpServletRequest request) {
+		String userIdRequestParameter = request.getParameter(AuthorizationConstants.USER_ID_PARAM);
+		return userIdRequestParameter == null ||
+				AuthorizationConstants.BOOTSTRAP_PRINCIPAL.ANONYMOUS_USER.getPrincipalId()
+					.equals(Long.parseLong(userIdRequestParameter));
+	}
+
+	@Override
+	public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+			throws Exception {
+		
+		// anonymous requests do not need to have scope checked, they have the same 
+		// access that unauthenticated requests have
+		if (isAnonymous(request)) {
+			return true;
+		}
+		
+		Set<String> missingScopes = new TreeSet<String>();
+		{
+			// if no scopes are specified by an annotation on the method, then we use these defaults
+			Set<OAuthScope> requiredScopes = new TreeSet<OAuthScope>(DEFAULT_SCOPES);
+			
+			if (!(handler instanceof HandlerMethod)) {
+				throw new IllegalStateException("Ths HandlerInterceptor should only be applied to HandlerMethods, but this handler is a "+handler.getClass());
+			}
+			
+			HandlerMethod handlerMethod = (HandlerMethod) handler;
+
+			// if no 'userId' parameter or access token header then this is 
+			// not an authenticated request, and no scope is required
+			if (!hasUserIdParameterOrAccessTokenHeader(handlerMethod)) {
+				return true;
+			}
+
+			RequiredScope requiredScopeAnnotation = handlerMethod.getMethodAnnotation(RequiredScope.class);
+			if (requiredScopeAnnotation != null) {
+				requiredScopes = new HashSet<OAuthScope>(Arrays.asList(requiredScopeAnnotation.value()));
+			}
+
+			List<OAuthScope> requestScopes = Collections.EMPTY_LIST;
+			String synapseAuthorizationHeader = request.getHeader(SYNAPSE_AUTHORIZATION_HEADER_NAME);
+			String accessToken = HttpAuthUtil.getBearerTokenFromAuthorizationHeader(synapseAuthorizationHeader);
+			if (accessToken!=null) {
+				Jwt<JwsHeader, Claims> jwt = oidcTokenHelper.parseJWT(accessToken);
+				requestScopes = ClaimsJsonUtil.getScopeFromClaims(jwt.getBody());
+			}
+
+			requiredScopes.removeAll(requestScopes);
+			for (OAuthScope scope: requiredScopes) {
+				missingScopes.add(scope.name());
```

Comment:
We can compute the missing scopes when needed to build the error message and we can check on the requiredScopes.isEmpty() instead at line 121